### ISSUE TITLE
Build of module desktop in branch release/v6.3.0 fails with link error on Ubuntu 18.04

### ISSUE THREAD

**Issue by h8ghbzc9bbr8c7** [ONLYOFFICE/build_tools#311]
The build of module desktop in branch release/v6.3.0 fails with link error on Ubuntu 18.04. The sequenve leading to this is

- git clone <build_tools>
- git checkout release/v6.3.0 in ./build_tools
- cd tools/linux in ./build_tools
- ./automate.py desktop

The build runs without errors until the link of DesktopEditors.
The log file from this last step is attached. The full log may be provided on request, it is large.

build_tools version: c89a54ccf47c4f72d6356d08613b1462a699b3b3
Operating System: Ubuntu 18.04
Compiler Version: gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)

[oo-build-error.log](https://github.com/ONLYOFFICE/build_tools/files/6480898/oo-build-error.log)

**Comment by ShockwaveNN** [comment]
Could you try to run just `./automate.py` on last step? This will be all products, but It's not much longer that only `desktop` build?

**Comment by h8ghbzc9bbr8c7** [comment]
I tried and unfortunately the result is the same - link error of DesktopEditors. May you share which distro you use?

**Comment by ShockwaveNN** [comment]
I've just recked using default Dockerfile included in project - it's using `FROM ubuntu:14.04` - try to use this repo

**Comment by h8ghbzc9bbr8c7** [comment]
Thanks again. I checked and this works. But this is not really helpful. Ubuntu 14.04 is old and unsupported since years. I can not even install this on a test server. So I repeat my suggestion in #310 to make the build_tools more functional to allow for broader testing and better feedback.

**Comment by ShockwaveNN** [comment]
> Ubuntu 14.04 is old and unsupported since years.

Yes you're right, but we got some clients who use system based on ubuntu 14.04
And since our build process build one generic linux exe for all supported debian-based system - we simply currently cannot update it to 16.04 for example and loose 14.04 support

But I'll take a look and send our developers info - crashing on ubuntu 16.04 is not normal

**Comment by ShockwaveNN** [comment]
@h8ghbzc9bbr8c7 I can confirm your error, talking with our dev team how we can resolve

I reproduced it simply by using default docker of this project and changing first line to `FROM ubuntu:18.04`

```
g++ -Wl,-rpath,'$ORIGIN' -Wl,-rpath,'$ORIGIN/system' -Wl,--disable-new-dtags -Wl,-rpath,'$ORIGIN' -Wl,-rpath,'$ORIGIN/converter' -static-libstdc++ -static-libgcc -Wl,-O1 -Wl,-rpath,/build_tools/tools/linux/qt_build/Qt-5.9.9/gcc_64/lib -Wl,-rpath-link,/build_tools/tools/linux/qt_build/Qt-5.9.9/gcc_64/lib -o build/linux_64/DesktopEditors build/linux_64/obj/csplash.o build/linux_64/obj/main.o build/linux_64/obj/asctabwidget.o build/linux_64/obj/cdownloadwidget.o build/linux_64/obj/cpushbutton.o build/linux_64/obj/cfiledialog.o build/linux_64/obj/cprintprogress.o build/linux_64/obj/ccefeventstransformer.o build/linux_64/obj/cascapplicationmanagerwrapper.o build/linux_64/obj/ctabbar.o build/linux_64/obj/casctabdata.o build/linux_64/obj/cstyletweaks.o build/linux_64/obj/chelp.o build/linux_64/obj/cmainpanel.o build/linux_64/obj/cmessage.o build/linux_64/obj/cfilechecker.o build/linux_64/obj/clogger.o build/linux_64/obj/clangater.o build/linux_64/obj/canimatedicon.o build/linux_64/obj/cscalingwrapper.o build/linux_64/obj/ctabundockevent.o build/linux_64/obj/cmainwindowbase.o build/linux_64/obj/ctabpanel.o build/linux_64/obj/csinglewindowbase.o build/linux_64/obj/ceditorwindow.o build/linux_64/obj/ccefeventsgate.o build/linux_64/obj/ceditortools.o build/linux_64/obj/ceventdriver.o build/linux_64/obj/csvgpushbutton.o build/linux_64/obj/cappupdater.o build/linux_64/obj/cappeventfilter.o build/linux_64/obj/cthemes.o build/linux_64/obj/cmainwindow.o build/linux_64/obj/cx11decoration.o build/linux_64/obj/gtk_addon.o build/linux_64/obj/cx11caption.o build/linux_64/obj/csinglewindow.o build/linux_64/obj/csinglewindowplatform.o build/linux_64/obj/singleapplication.o build/linux_64/obj/cdialogopenssl.o build/linux_64/obj/cdialogcertificateinfo.o build/linux_64/obj/cmainpanelimpl.o build/linux_64/obj/utils.o build/linux_64/obj/qrc_resources.o build/linux_64/obj/moc_asctabwidget.o build/linux_64/obj/moc_cdownloadwidget.o build/linux_64/obj/moc_cpushbutton.o build/linux_64/obj/moc_cfiledialog.o build/linux_64/obj/moc_cprintprogress.o build/linux_64/obj/moc_cascapplicationmanagerwrapper.o build/linux_64/obj/moc_ctabbar.o build/linux_64/obj/moc_cmainpanel.o build/linux_64/obj/moc_cfilechecker.o build/linux_64/obj/moc_clogger.o build/linux_64/obj/moc_clangater.o build/linux_64/obj/moc_ctabpanel.o build/linux_64/obj/moc_ccefeventsgate.o build/linux_64/obj/moc_ceventdriver.o build/linux_64/obj/moc_cappeventfilter.o build/linux_64/obj/moc_cmainwindow.o build/linux_64/obj/moc_cx11decoration.o build/linux_64/obj/moc_csinglewindow.o build/linux_64/obj/moc_singleapplication.o build/linux_64/obj/moc_cdialogopenssl.o   -L/desktop-apps/win-linux/../../core/Common/3dParty/cef/linux_64/build -lcef -L/desktop-apps/win-linux/../../core/build/lib/linux_64 -lPdfReader -lPdfWriter -lDjVuFile -lXpsFile -lHtmlRenderer -lUnicodeConverter -lhunspell -looxmlsignature -lkernel -lgraphics -lvideoplayer -lascdocumentscore -lqtascdocumentscore -lX11 /desktop-apps/win-linux/../../core/Common/3dParty/icu/linux_64/build/libicuuc.so.58 /desktop-apps/win-linux/../../core/Common/3dParty/icu/linux_64/build/libicudata.so.58 -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -latk-1.0 -lgobject-2.0 -L/build_tools/tools/linux/qt_build/Qt-5.9.9/gcc_64/lib -lQt5PrintSupport -lQt5MultimediaWidgets -lpulse-mainloop-glib -lpulse -lglib-2.0 -lQt5Svg -lQt5Widgets -lQt5Multimedia -lQt5X11Extras -lQt5Gui -lQt5Network -lQt5Core -lGL -lpthread 
build/linux_64/obj/cappupdater.o: In function `(anonymous namespace)::CThreadProc::run()':
cappupdater.cpp:(.text+0x860): warning: the use of `tmpnam' is dangerous, better use `mkstemp'
build/linux_64/obj/cdialogopenssl.o: In function `CDialogOpenSsl::checkCertificate()':
cdialogopenssl.cpp:(.text+0xdb7): undefined reference to `NSOpenSSL::LoadCert(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::string)'
cdialogopenssl.cpp:(.text+0x1227): undefined reference to `NSOpenSSL::LoadKey(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::string)'
build/linux_64/obj/utils.o: In function `InputArgs::init(int, char**)':
utils.cpp:(.text+0x220b): undefined reference to `NSFile::CUtf8Converter::GetUnicodeStringFromUTF8(unsigned char*, long)'
build/linux_64/obj/main.o: In function `main':
main.cpp:(.text.startup+0x9e5): undefined reference to `CAscApplicationSettings::SetUserDataPath(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >)'
main.cpp:(.text.startup+0xa0d): undefined reference to `NSFile::GetProcessDirectory()'
build/linux_64/obj/asctabwidget.o: In function `CAscTabWidget::addEditor(COpenOptions const&)':
asctabwidget.cpp:(.text+0xc2d): undefined reference to `CCefView::load(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
asctabwidget.cpp:(.text+0xdb5): undefined reference to `CCefViewEditor::GetFileFormat(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/asctabwidget.o: In function `CAscTabWidget::addPortal(QString const&, QString const&, QString const&, QString const&)':
asctabwidget.cpp:(.text+0x14a7): undefined reference to `CCefView::SetExternalCloud(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
asctabwidget.cpp:(.text+0x151e): undefined reference to `CCefView::load(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/asctabwidget.o: In function `CAscTabWidget::addOAuthPortal(QString const&, QString const&, QString const&, QString const&)':
asctabwidget.cpp:(.text+0x1ab3): undefined reference to `CCefView::load(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
asctabwidget.cpp:(.text+0x1bdd): undefined reference to `CCefView::SetExternalCloud(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
asctabwidget.cpp:(.text+0x1c98): undefined reference to `CCefView::load(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/asctabwidget.o: In function `CAscTabWidget::tabIndexByUrl(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)':
asctabwidget.cpp:(.text+0x2571): undefined reference to `CAscApplicationManager::GetViewByUrl(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/asctabwidget.o: In function `CAscTabWidget::updatePortal(int, QString const&)':
asctabwidget.cpp:(.text+0x28f5): undefined reference to `CCefView::load(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
asctabwidget.cpp:(.text+0x2937): undefined reference to `CCefView::load(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/cascapplicationmanagerwrapper.o: In function `CAscApplicationManagerWrapper::Logout(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)':
cascapplicationmanagerwrapper.cpp:(.text+0x2084): undefined reference to `CAscApplicationManager::Logout(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >)'
build/linux_64/obj/cascapplicationmanagerwrapper.o: In function `CAscApplicationManagerWrapper::applyTheme(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, bool)':
cascapplicationmanagerwrapper.cpp:(.text+0x27ef): undefined reference to `CAscApplicationManager::InitAdditionalEditorParams(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >&)'
build/linux_64/obj/cascapplicationmanagerwrapper.o: In function `CAscApplicationManagerWrapper::userSettings(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)':
cascapplicationmanagerwrapper.cpp:(.text+0x3148): undefined reference to `CUserSettings::Get(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/cascapplicationmanagerwrapper.o: In function `CAscApplicationManagerWrapper::setUserSettings(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)':
cascapplicationmanagerwrapper.cpp:(.text+0x31d8): undefined reference to `CUserSettings::Set(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/cascapplicationmanagerwrapper.o: In function `CAscApplicationManagerWrapper::initializeApp()':
cascapplicationmanagerwrapper.cpp:(.text+0x4ce3): undefined reference to `CAscApplicationManager::InitAdditionalEditorParams(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >&)'
build/linux_64/obj/cascapplicationmanagerwrapper.o: In function `CAscApplicationManagerWrapper::applySettings(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)':
cascapplicationmanagerwrapper.cpp:(.text+0x5791): undefined reference to `CAscApplicationManager::InitAdditionalEditorParams(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >&)'
build/linux_64/obj/cascapplicationmanagerwrapper.o: In function `CAscApplicationManagerWrapper::processCommonEvent(NSEditorApi::CAscCefMenuEvent*)':
cascapplicationmanagerwrapper.cpp:(.text+0x7f4c): undefined reference to `CAscApplicationManager::GetViewByUrl(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/cascapplicationmanagerwrapper.o: In function `prepareMainWindow()':
cascapplicationmanagerwrapper.cpp:(.text+0x8e04): undefined reference to `CCefView::load(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/cascapplicationmanagerwrapper.o: In function `CAscApplicationManagerWrapper::handleInputCmd(std::vector<std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::allocator<std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > > > const&)':
cascapplicationmanagerwrapper.cpp:(.text+0x9593): undefined reference to `NSFile::CUtf8Converter::GetUtf8StringFromUnicode2(wchar_t const*, long, bool)'
build/linux_64/obj/ctabpanel.o: In function `CTabPanel::openLocalFile(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, int, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)':
ctabpanel.cpp:(.text+0x4a8): undefined reference to `CCefViewEditor::OpenLocalFile(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, int const&, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/ctabpanel.o: In function `CTabPanel::openLocalFile(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)':
ctabpanel.cpp:(.text+0x4e5): undefined reference to `CCefViewEditor::GetFileFormat(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
ctabpanel.cpp:(.text+0x50e): undefined reference to `CCefViewEditor::OpenLocalFile(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, int const&, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/ctabpanel.o: In function `CTabPanel::createLocalFile(int, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)':
ctabpanel.cpp:(.text+0x561): undefined reference to `CCefViewEditor::CreateLocalFile(int const&, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/ceditortools.o: In function `CEditorTools::createEditorPanel(COpenOptions const&, QRect const&)':
ceditortools.cpp:(.text+0x19d): undefined reference to `CCefView::load(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/ceditortools.o: In function `CEditorTools::processLocalFileSaveAs(NSEditorApi::CAscCefMenuEvent const*)':
ceditortools.cpp:(.text+0x171f): undefined reference to `CAscApplicationManager::GetFileFormatByExtentionForSave(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/cappupdater.o: In function `(anonymous namespace)::CThreadProc::run()':
cappupdater.cpp:(.text+0x827): undefined reference to `CFileDownloader::CFileDownloader(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, bool)'
cappupdater.cpp:(.text+0x87f): undefined reference to `NSFile::CUtf8Converter::GetUnicodeStringFromUTF8(unsigned char*, long)'
cappupdater.cpp:(.text+0x88a): undefined reference to `CFileDownloader::SetFilePath(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
cappupdater.cpp:(.text+0x8c7): undefined reference to `CFileDownloader::GetFilePath()'
build/linux_64/obj/cappupdater.o: In function `CAppUpdater::parse_app_cast(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)':
cappupdater.cpp:(.text+0xd18): undefined reference to `NSFile::CUtf8Converter::GetUtf8StringFromUnicode(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
cappupdater.cpp:(.text+0x1084): undefined reference to `NSFile::CFileBinary::Remove(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&)'
build/linux_64/obj/moc_cascapplicationmanagerwrapper.o:(.data.rel.ro._ZTV29CAscApplicationManagerWrapper[_ZTV29CAscApplicationManagerWrapper]+0xc0): undefined reference to `CAscApplicationManager::EndSaveDialog(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, unsigned int)'
collect2: error: ld returned 1 exit status
/build_tools/../desktop-apps/win-linux/Makefile.ASCDocumentEditorlinux_64:422: recipe for target 'build/linux_64/DesktopEditors' failed
make[1]: *** [build/linux_64/DesktopEditors] Error 1
make[1]: Leaving directory '/desktop-apps/win-linux'
makefiles/build.makefile_linux_64:1597: recipe for target 'sub--desktop-apps-win-linux-ASCDocumentEditor-pro-make_first-ordered' failed
make: *** [sub--desktop-apps-win-linux-ASCDocumentEditor-pro-make_first-ordered] Error 2
Error (make): 2
Error (./make.py): 1
```

**Comment by h8ghbzc9bbr8c7** [comment]
Thank you very much for your efforts and your support!

**Comment by ShockwaveNN** [comment]
@h8ghbzc9bbr8c7 Hi again, our developers fixed a problem with Ubuntu on 18.04 with [this](https://github.com/ONLYOFFICE/desktop-apps/commit/962c27ac03b897d35d93a36a2b936305cfda72ca) commit

You can try to build from branch hotfix/v6.3.1 to see result

**Comment by h8ghbzc9bbr8c7** [comment]
Works perfectly, works perfectly (now I will look for bugs :-) )
